File: cryptokit/Sources/CryptoKitSrc/hkdf.swift
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import CryptoKit
import Foundation

@_cdecl("extractHKDF")
public func extractHKDF(
    hashFunction: Int32,
    secretPointer: UnsafePointer<UInt8>, secretLength: Int,
    saltPointer: UnsafePointer<UInt8>, saltLength: Int,
    prkPointer: UnsafeMutablePointer<UInt8>, prkLength: Int
) -> Int {
    // Convert input pointers to Data
    let secretData: SymmetricKey = SymmetricKey(data: Data(bytes: secretPointer, count: secretLength))
    let saltData = Data(bytes: saltPointer, count: saltLength)

    let prk: any MessageAuthenticationCode

    switch hashFunction {
    case 1:
        prk = HKDF<Insecure.SHA1>.extract(
            inputKeyMaterial: secretData,
            salt: saltData
        )
    case 2:
        prk = HKDF<SHA256>.extract(
            inputKeyMaterial: secretData,
            salt: saltData
        )
    case 3:
        prk = HKDF<SHA384>.extract(
            inputKeyMaterial: secretData,
            salt: saltData
        )
    case 4:
        prk = HKDF<SHA512>.extract(
            inputKeyMaterial: secretData,
            salt: saltData
        )
    default:
        return -1 // Unsupported hash function
    }

    // Convert prk to Data
    let prkData = prk.withUnsafeBytes { Data($0) }

    prkData.copyBytes(to: prkPointer, count: prkData.count)

    return 0
}

@_cdecl("expandHKDF")
public func expandHKDF(
    hashFunction: Int32,
    prkPointer: UnsafePointer<UInt8>, prkLength: Int,
    infoPointer: UnsafePointer<UInt8>, infoLength: Int,
    derivedKeyPointer: UnsafeMutablePointer<UInt8>, derivedKeyLength: Int
) -> Int {
    // Convert input pointers to Data
    let prkData: SymmetricKey = SymmetricKey(data: Data(bytes: prkPointer, count: prkLength))
    let infoData = Data(bytes: infoPointer, count: infoLength)

    let derivedKey: SymmetricKey

    switch hashFunction {
    case 1:
        derivedKey = HKDF<Insecure.SHA1>.expand(
            pseudoRandomKey: prkData,
            info: infoData,
            outputByteCount: derivedKeyLength
        )
    case 2:
        derivedKey = HKDF<SHA256>.expand(
            pseudoRandomKey: prkData,
            info: infoData,
            outputByteCount: derivedKeyLength
        )
    case 3:
        derivedKey = HKDF<SHA384>.expand(
            pseudoRandomKey: prkData,
            info: infoData,
            outputByteCount: derivedKeyLength
        )
    case 4:
        derivedKey = HKDF<SHA512>.expand(
            pseudoRandomKey: prkData,
            info: infoData,
            outputByteCount: derivedKeyLength
        )
    default:
        return -1 // Unsupported hash function
    }

    // Convert derivedKey to Data
    let derivedKeyData = derivedKey.withUnsafeBytes { Data($0) }

    derivedKeyData.copyBytes(to: derivedKeyPointer, count: derivedKeyData.count)

    return 0
}

File: commoncrypto/hkdf.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

//go:build darwin

package commoncrypto

import (
	"hash"

	"github.com/microsoft/go-crypto-darwin/cryptokit"
)

// ExtractHKDF performs the extract step of HKDF using the specified hash function.
func ExtractHKDF(h func() hash.Hash, secret, salt []byte) ([]byte, error) {
	hash, err := hashToCryptoHash(h())
	if err != nil {
		return nil, err
	}

	prk, err := cryptokit.ExtractHKDF(hash, secret, salt)
	if err != nil {
		return nil, err
	}

	return prk, nil
}

// ExpandHKDF performs the expand step of HKDF using the specified hash function.
func ExpandHKDF(h func() hash.Hash, pseudorandomKey, info []byte, keyLength int) ([]byte, error) {
	hash, err := hashToCryptoHash(h())
	if err != nil {
		return nil, err
	}

	expandedKey, err := cryptokit.ExpandHKDF(hash, pseudorandomKey, info, keyLength)
	if err != nil {
		return nil, err
	}

	return expandedKey, nil
}
